Added linalg.slogdet

IvanYashchuk opened the pull request:
This PR adds `torch.linalg.slogdet`.

Changes compared to the original torch.slogdet:

- Complex input now works as in NumPy
- Added out= variant (allocates temporary and makes a copy for now)
- Updated `slogdet_backward` to work with complex input

Ref. #42666

IvanYashchuk committed:
wip linalg.slogdet

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-slogdet

IvanYashchuk committed:
Removed duplicated tests

IvanYashchuk committed:
Added linalg.slogdet to tests

IvanYashchuk committed:
Implemented complex input support; added tests

IvanYashchuk committed:
Updated docs

IvanYashchuk committed:
Added out= tests

IvanYashchuk committed:
wip started OpInfo-based test; doesn't work yet

IvanYashchuk committed:
Updated slogdet_backward to work for complex input

IvanYashchuk committed:
Fix test_unsupported_dtypes

IvanYashchuk committed:
Skip test_variant_consistency_jit

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-slogdet

IvanYashchuk committed:
flake8

IvanYashchuk committed:
Updated comment in tests

IvanYashchuk committed:
Added entry to test_namedtuple_return_api.py

mruberry commented:
This test failure looks real:

```
test_slogdet_cpu_float32 - TestLinalgCPU
test_linalg.py

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 273, in instantiated_test
    result = test_fn(self, *args)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 545, in dep_fn
    return fn(slf, device, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 545, in dep_fn
    return fn(slf, device, *args, **kwargs)
  File "test_linalg.py", line 4726, in test_slogdet
    run_test(matsize, batchdims, mat_chars=['singular'])
  File "test_linalg.py", line 4712, in run_test
    self.assertEqual(expected_value[0], actual_value[0], atol=self.precision, rtol=self.precision)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1187, in assertEqual
    exact_dtype=exact_dtype, exact_device=exact_device)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1068, in assertEqual
    exact_dtype=exact_dtype, exact_device=exact_device)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1198, in assertEqual
    super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
AssertionError: False is not true : Scalars failed to compare as equal! Comparing 1.0 and 0.0 gives a difference of 1.0, but the allowed difference with rtol=0.001 and atol=0.001 is only 0.001!
```

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -80,21 +80,46 @@ Tensor logdet(const Tensor& self) {
   squareCheckInputs(self);
-  TORCH_CHECK((at::isFloatingType(self.scalar_type()) || at::isComplexType(self.scalar_type())),
-              "Expected a floating point tensor as input");
+  TORCH_CHECK((at::isFloatingType(self.scalar_type()) || at::isComplexType(self.scalar_type())) && self.dim() >= 2,
```
Does this handle bfloat16 and float16 inputs, too?

IvanYashchuk replied:
No, this was copied from the old code of `slogdet` and I didn't check what `isFloatingType` does.
I'll change it to accept only float, double, cfloat, cdouble.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -80,21 +80,46 @@ Tensor logdet(const Tensor& self) {
-              "Expected a floating point tensor as input");
+  TORCH_CHECK((at::isFloatingType(self.scalar_type()) || at::isComplexType(self.scalar_type())) && self.dim() >= 2,
+              "linalg_slogdet(", self.scalar_type(), "{", self.sizes(), "}): expected a tensor with 2 or more dimensions "
+              "of floating types");
```
Not of floating types - floating point or complex tensors are OK

mruberry commented on test/test_linalg.py:
```diff
@@ -4676,6 +4676,58 @@ def run_test(p):
+    @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
+    @precisionOverride({torch.float32: 1e-3, torch.complex64: 1e-3,
+                        torch.float64: 1e-8, torch.complex128: 1e-8})
+    def test_slogdet(self, device, dtype):
```
Are there are also test cases for the assertions this may throw?

IvanYashchuk replied:
I added `test_slogdet_errors_and_warnings`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -88,6 +88,41 @@
+slogdet = _add_docstr(_linalg.linalg_slogdet, r"""
+linalg.slogdet(input) -> (Tensor, Tensor)
+
+Calculates the sign and log absolute value of the determinant(s) of a square matrix or
```
"and log absolute value" -> "and natural logarithmic of the absolute values of the determinants of..."?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -88,6 +88,41 @@
+
+.. note:: When given inputs on a CUDA device, this function synchronizes that device with the CPU.
+
+.. note:: If ``input`` has zero determinant, this returns ``(0, -inf)``.
```
This should describe what happens in the batched case, too. If `input` is reference in isolation it should be ":attr:\`input\`".

IvanYashchuk replied:
Changed to 
```python
.. note:: For matrices that have zero determinant, this returns ``(0, -inf)``.
          If :attr:`input` is batched then the entries in the result tensors corresponding to matrices with
          the zero determinant have sign 0 and the natural logarithm of the absolute value of the determinant -inf.
```

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -88,6 +88,41 @@
+                    where `*` is one or more batch dimensions.
+
+Returns:
+    A namedtuple (sign, logabsdet) containing the sign of the determinant, and the log
```
See logarithm description above (which log? clarify "absolute determinant")

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -281,6 +281,31 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad):
+    from torch.testing._internal.common_utils import random_hermitian_matrix, random_hermitian_pd_matrix
+
+    # tests cases are copied from 'method_tests'
+    test_inputs = (
```
Does this handle empty tensors? That is, where one or more of the batch dimensions is zero, or where one of the matrices' dimensions is zero?

IvanYashchuk replied:
`torch.linalg.slogdet` works for such inputs and I added them to the tests.

mruberry reviewed (commented):
Hey @IvanYashchuk!

Overall this looks really good as usual. Just a few questions/comments to work through.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-slogdet

IvanYashchuk committed:
Fixed supported dtypes check

IvanYashchuk committed:
Added test_slogdet_errors_and_warnings

IvanYashchuk committed:
Added 0x0, 0xSxS, 3x3x0x0 to sample_inputs_slogdet

IvanYashchuk committed:
Added 0x0 and zero batch dimension test cases

IvanYashchuk committed:
Removed todo comment

IvanYashchuk committed:
Updated documentation

IvanYashchuk committed:
Removed scaling

IvanYashchuk committed:
Merge branch 'master' into linalg-slogdet

IvanYashchuk committed:
Merge branch 'master' into linalg-slogdet

IvanYashchuk committed:
Fixed c++ docs check

AssertionError: linalg_slogdet.sign takes out arguments and has to be written in the new style. Please add `use_c10_dispatcher: full` to your operator in native_functions.yaml and write the C++ implementation to take out arguments in the end.

IvanYashchuk committed:
Added same device check

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-slogdet

IvanYashchuk committed:
Fix CI failure: Torch not compiled with CUDA enabled

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-slogdet

IvanYashchuk committed:
Merge branch 'master' into linalg-slogdet

IvanYashchuk committed:
Merge branch 'master' into linalg-slogdet

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1567,6 +1567,19 @@ def random_symmetric_psd_matrix(l, *batches, **kwargs):
+    The following example creates a tensor of size 2 x 4 x 3 x 3
+    >>> matrices = random_hermitian_psd_matrix(3, 2, 4, dtype=dtype, device=device)
+    """
+    dtype = kwargs.get('dtype', torch.double)
```
Let's make these required keyword-only arguments for readability.

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1541,7 +1541,7 @@ def random_square_matrix_of_rank(l, rank, dtype=torch.double, device='cpu'):
         elif s[i] == 0:
             s[i] = 1
-    return u.mm(torch.diag(s)).mm(v.transpose(0, 1))
+    return u.mm(torch.diag(s).to(dtype)).mm(v.transpose(0, 1))
```
It's unfortunate that `torch.diag` doesn't take tensor options, despite it being a tensor creation op.

IvanYashchuk replied:
It doesn't need to be transferred, the device of `diag(s)` is `s.device()`.
`to(dtype)` is needed here because for cfloat and cdouble input, `u`and`v` are of type cfloat/cdouble while `s` is float/double to be able to multiply them `s` needs to be transformed into cfloat/cdouble. Matrix multiplication doesn't allow mixing real and complex arguments.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1002,6 +1030,19 @@ def sample_inputs_pinverse(op_info, device, dtype, requires_grad=False):
                                 dtypes=[torch.bfloat16])),
                    promotes_integers_to_float=True,
                    handles_complex_extremals=False),
+    OpInfo('linalg.slogdet',
```
How time consuming are the grad and gradgrad checks for slogdet on CPU and CUDA?

IvanYashchuk replied:
Before this commit b3f8932c27:
```
39.03s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_linalg_slogdet_cuda_complex128
14.70s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_linalg_slogdet_cpu_complex128
10.48s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_linalg_slogdet_cuda_float64
4.61s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_slogdet_cuda_complex128
3.81s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_linalg_slogdet_cpu_float64
2.33s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_linalg_slogdet_cuda_complex128
2.12s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_slogdet_cuda_float64
1.72s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_linalg_slogdet_cpu_complex128
0.75s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_linalg_slogdet_cpu_float64
```
and after:
```
4.84s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_linalg_slogdet_cuda_complex128
1.91s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_linalg_slogdet_cuda_complex128
1.88s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_linalg_slogdet_cpu_complex128
1.54s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_slogdet_cuda_complex128
1.33s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_linalg_slogdet_cuda_float64
0.51s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_linalg_slogdet_cpu_float64
0.39s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_slogdet_cuda_float64
0.22s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_linalg_slogdet_cpu_complex128
```

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -287,6 +287,34 @@ def sample_inputs_tensor_split(op_info, device, dtype, requires_grad):
+    from torch.testing._internal.common_utils import random_hermitian_matrix, random_hermitian_pd_matrix
+
+    # tests cases are copied from 'method_tests'
+    # but make_nonzero_det is not used, since there is no negative or positive determinant
```
Additional cases could be added if the dtype is not complex, though. Would that make sense here?

IvanYashchuk replied:
I don't think it is necessary, there is no special handling of the sign value in the code so no need to test specifically the positive or negative signs. In fact, since slogdet_backward is SVD-based we shouldn't have too many test inputs here otherwise it gets expensive very quickly.
I removed a few test cases in b3f8932 which I think are not needed making tests much faster.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -88,6 +88,43 @@
 
+slogdet = _add_docstr(_linalg.linalg_slogdet, r"""
+linalg.slogdet(input) -> (Tensor, Tensor)
+
```
"of the absolute values of the determinants of the determinant(s) of a square matrix or batches of square matrices :attr:`input`." -> "of the absolute value of a square matrix's determinant, or of the absolute values of the determinants of a batch of square matrices :attr\`input\`."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -88,6 +88,43 @@
+Calculates the sign and natural logarithm of the absolute values of the determinants of the determinant(s)
+of a square matrix or batches of square matrices :attr:`input`.
+
+Supports input of ``float``, ``double``, ``cfloat`` and ``cdouble`` datatypes.
```
Let's ditch the formatting around the dtypes so they render as regular words

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -88,6 +88,43 @@
+
+.. note:: For matrices that have zero determinant, this returns ``(0, -inf)``.
+          If :attr:`input` is batched then the entries in the result tensors corresponding to matrices with
+          the zero determinant have sign 0 and the natural logarithm of the absolute value of the determinant -inf.
```
"the zero determinant" -> "a zero determinant"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -88,6 +88,43 @@
+
+.. note:: When given inputs on a CUDA device, this function synchronizes that device with the CPU.
+
+.. note:: For matrices that have zero determinant, this returns ``(0, -inf)``.
```
Could we simplify this by saying, "Matrices with zero determinants are treated as having a sign of zero, even though the natural logarithm of zero is treated as negative infinity."?

IvanYashchuk replied:
I think it makes it more complicated. This note together with "sign * exp(logabsdet)" is simple?

mruberry replied:
OK, if you think this is simpler let's stick with it.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -88,6 +88,43 @@
+                    where `*` is one or more batch dimensions.
+
+Returns:
+    A namedtuple (sign, logabsdet) containing the sign of the determinant, and the natural logarithm
```
"determinant, and" -> "determinant and"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -88,6 +88,43 @@
+
+Returns:
+    A namedtuple (sign, logabsdet) containing the sign of the determinant, and the natural logarithm
+    of the absolute value of determinant.
```
"of determinant." -> "of the determinant, respectively."

mruberry commented on test/test_ops.py:
```diff
@@ -84,14 +84,19 @@ def variant_out_fn(*args, **kwargs):
             else:
                 variant_out_fn = variant
             partial_fn = partial(variant_out_fn, **sample.kwargs)
```
Since partial_fn lives entirely within `fn` now I think its definition can be moved inside it, too

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -88,6 +88,43 @@
+                    where `*` is one or more batch dimensions.
+
+Returns:
+    A namedtuple (sign, logabsdet) containing the sign of the determinant, and the natural logarithm
```
`logabsdet` is a more accurate name than `logdet`, but `logdet` is the name NumPy uses. 

As much as I prefer `logabsdet`, seems like we should use `logdet` for consistency?

IvanYashchuk replied:
NumPy doesn't use named tuples and `logdet` is only in its documentation, which could be improved. `logabsdet` is more accurate and I think we should keep it.

mruberry replied:
Good point; I agree with you.

mruberry commented on test/test_namedtuple_return_api.py:
```diff
@@ -65,6 +65,7 @@ def test_namedtuple_return(self):
             op(operators=['triangular_solve'], input=(a,), names=('solution', 'cloned_coefficient'), hasout=True),
             op(operators=['lstsq'], input=(a,), names=('solution', 'QR'), hasout=True),
             op(operators=['linalg_eigh'], input=("L",), names=('eigenvalues', 'eigenvectors'), hasout=True),
+            op(operators=['linalg_slogdet'], input=(), names=('sign', 'logabsdet'), hasout=True),
```
See `logabset` vs `logdet` question below

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -9961,6 +9961,19 @@
+  use_c10_dispatcher: full
+  variants: function
+  dispatch:
+    DefaultBackend: linalg_slogdet
```
We should probably just use CPU, CUDA instead of DefaultBackend here?

IvanYashchuk replied:
Yep, we can start using `CPU, CUDA`. Many other linalg function use DefaultBackend now and I followed the pattern here.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -80,21 +80,51 @@ Tensor logdet(const Tensor& self) {
-  TORCH_CHECK((at::isFloatingType(self.scalar_type()) || at::isComplexType(self.scalar_type())),
-              "Expected a floating point tensor as input");
+  ScalarType t = self.scalar_type();
+  TORCH_CHECK((t == ScalarType::Double || t == ScalarType::Float || t == ScalarType::ComplexFloat || t == ScalarType::ComplexDouble)
```
We should probably separate these checks and use the pattern we've been working with so far of:

function_name expected ... but got ...

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -80,21 +80,51 @@ Tensor logdet(const Tensor& self) {
   // U is singular when U(i, i) = 0 for some i in [1, self.size(-1)].
   // Since abslogdet_val cannot take nan, no special case handling is required.
-  auto abslogdet_val = diag_U.abs_().log_().sum(-1);
+  // in-place abs is not supported for complex tensors
```
What about diverging the code path based on dtype to perform the operation inplace for floating types and out-of-place for complex types?

IvanYashchuk replied:
Done in b8ea890

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -88,6 +88,43 @@
+Returns:
+    A namedtuple (sign, logabsdet) containing the sign of the determinant, and the natural logarithm
+    of the absolute value of determinant.
+
```
One thing that the NumPy documentation briefly mentions but is absent here is the interpretation of sign. Maybe a note on how to interpret the sign value would be interesting?

IvanYashchuk replied:
Is a sentence "The determinant can be computed with ``sign * exp(logabsdet)``." enough? I think having it we don't need to expand further what sign means and what values can it have for complex input or zero determinant matrices.

mruberry replied:
OK, sounds good.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -80,21 +80,51 @@ Tensor logdet(const Tensor& self) {
+  return std::tuple<Tensor&, Tensor&>(sign, logabsdet);
+}
+
+std::tuple<Tensor, Tensor> slogdet(const Tensor& self) {
```
So with this change `torch.slogdet` is an alias for `torch.linalg.slogdet`? 

Although they might diverge if we change the namedtuple names to match NumPy.

IvanYashchuk replied:
Yes, now `torch.slogdet` is an alias for `torch.linalg.slogdet` so it inherits for example complex support from it. Is there a special way to register aliases?

mruberry replied:
Kind of, but we don't need to worry about it at this time.

Would you file an issue, however, that torch.slogdet should be made an alias of torch.linalg.slogdet? Then we can follow-up on that (low priority) issue at a later tdate.

Aliasing is a pattern more than a registry, although there is a registry in the jit that remaps aliases, there's no actual value from remapping svd to linalg.svd in torchscript. That remapping is intended for more common operations, like chunk or split or div, which torchscript might be able to optimize.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1002,6 +1030,19 @@ def sample_inputs_pinverse(op_info, device, dtype, requires_grad=False):
+           skips=(
+               # These tests do not work with output_func=itemgetter(1)
+               # TODO: remove this once https://github.com/pytorch/pytorch/issues/49326 is resolved
+               SkipInfo('TestCommon', 'test_variant_consistency_jit'))),
```
This is missing a comma, which is causing the failing CI builds.

I always thought it was weird that (5) in Python is not a tuple.

mruberry reviewed (commented):
Hey @IvanYashchuk! As usual this looks very good. I have a few questions/comments inline. The tests look especially thorough, but a missing comma is causing build failures. Looking forward to an update!

IvanYashchuk committed:
Add missing comma

IvanYashchuk committed:
Changed kwargs of random_hermitian_psd_matrix

IvanYashchuk committed:
Added a ternary for abs_ vs abs

IvanYashchuk committed:
Changed DefaultBackend -> CPU, CUDA

IvanYashchuk committed:
Updated docs

IvanYashchuk committed:
self.dim >= 2 is alerady checked in squareCheckInputs

IvanYashchuk committed:
Make test_ops faster by changing sample_inputs_slogdet

IvanYashchuk committed:
partial is absorbed into 'fn'

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-slogdet

IvanYashchuk committed:
Removed unused imports

IvanYashchuk commented:
@mruberry, holidays are over and I've finally updated this pull request. It's ready for another round of review.

IvanYashchuk committed:
Merge branch 'master' into linalg-slogdet

mruberry commented:
Looks like this just needs a rebase. Also, I thought we had to register complex gradients somewhere? Did I miss that or is it not needed in this PR?

Sorry for the delay in reviewing this, by the way. It's been an incredibly busy week. Things should get back to normal next week.

mruberry reviewed (approved):
Awesome.

Sorry for the delay in reviewing this, @IvanYashchuk. It's been an incredibly busy week, and I appreciate your help reviewing cuSOLVER svd and lstsq. Things should be back to normal next week, but if you're enjoying helping with reviews it'd be great to have you review more often. Let me know.

On this PR, I thought supporting complex autograd required adding the function's name to a list somewhere -  did I miss that in this PR or does linalg.slogdet not need to do that for some reason? 

Finally, this just needs a rebase (edit: and let's remove some method_test entries).

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -968,6 +985,19 @@ def sample_inputs_fliplr_flipud(op_info, device, dtype, requires_grad):
+           skips=(
+               # These tests do not work with output_func=itemgetter(1)
+               # TODO: remove this once https://github.com/pytorch/pytorch/issues/49326 is resolved
+               SkipInfo('TestCommon', 'test_variant_consistency_jit'),)),
```
While we're here let's remove the slogdet entries in method_tests:

https://github.com/pytorch/pytorch/blob/4a3a37886cdbdca108536a2730c034db4f96f095/torch/testing/_internal/common_methods_invocations.py#L2295

slogdet is just a deprecated alias for linalg.slogdet at this point.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-slogdet

IvanYashchuk committed:
Removed the slogdet entries in method_tests

IvanYashchuk commented:
@mruberry I removed method_test entries.
linalg.slogdet does not need to be registered for complex autograd because this function maps complex-valued input to real-valued output and it's not differentiable wrt `sign` (which is complex). See https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_variable_type.py#L69-L70

> if you're enjoying helping with reviews it'd be great to have you review more often. Let me know.
Sure! Assign me as a reviewer wherever you think I can help

mruberry commented:
Import is needed:

```
Jan 15 01:13:12   File "test_ops.py", line 60, in <module>
Jan 15 01:13:12     _gradcheck_ops = partial(ops, dtypes=OpDTypes.supported,
Jan 15 01:13:12 NameError: name 'partial' is not defined
```

IvanYashchuk committed:
Get partial back

IvanYashchuk commented:
@mruberry I am sorry I overlooked that.

mruberry commented:
> @mruberry I am sorry I overlooked that.

No worries. Happens to everyone.
